handlers/smshandler/smshandler.go

```diff
@@ -76,9 +99,35 @@ func (sh *SmsHandler) Init(opts ...string) error {
 	if importFile != "" {
 		sh.importCSVfile(importFile)
 	}
+
+	// create SMS queue
+	sh.SmsThrottle = DefaultSMSthrottleTime
+	sh.smsQueue = newSmsQueue(
+		smsCoolDownTime,
+		sh.SendChallenge,
+	)
+	go sh.smsQueue.run()
+	go sh.smsQueueController()
 	return nil
 }
 
+// We add a wait in order to not stress the SMS provider API
+func (sh *SmsHandler) smsQueueController() {
+	for {
+		r := <-sh.smsQueue.response
+		if r.success {
+			if err := sh.stg.SetAttempts(r.userID, r.electionID, -1); err != nil {
+				log.Warnf("challenge cannot be sent: %v", err)
+			} else {
+				log.Infof("challenge successfully sent to %s", r.userID)
+			}
+		} else {
+			log.Infof("challenge sending failed for %s", r.userID)
+		}
+		time.Sleep(sh.SmsThrottle)
+	}
+}
```

Comment:
i'm believe this needs refactoring. the way it's written, SetAttempts is only called when the failure is reported, and in production (especially in overload conditions) this might take a while (several seconds at least). during this time window, let's say `MaxAttempts = 5`, the user could re-request infinite attempts until the first 5 report success.

i think it's safer to `SetAttempts(-1)` as soon as the sms is enqueued for delivery, and in case a failure is reported, increase it again `SetAttempts(+1)`.

so, there's no possibility whatsoever that more than MaxAttempts smses are "in flight" (enqueued, being retried, etc)

Reply:
also, implementing here the SmsThrottle sounds weird to me, i'd put it inside queue.go run()

Reply:
Thanks for looking deep into the details :)

> i'm believe this needs refactoring. the way it's written, SetAttempts is only called when the failure is reported, and in production (especially in overload conditions) this might take a while (several seconds at least). during this time window, let's say MaxAttempts = 5, the user could re-request infinite attempts until the first 5 report success.
i think it's safer to SetAttempts(-1) as soon as the sms is enqueued for delivery, and in case a failure is reported, increase it again SetAttempts(+1).

The previous approach was exactly this, but there is a problem: what happens if the server is shutdown or restarted on that moment? Its attempt would be already consumed but the challenge has not been actually sent.  In a congested scenario with the queue full, it can be quite a bad behavior.

A potential solution is to use a disk queue but I did not find anything good enough and we have not time for implementing that.

So I preferred to change the approach, only remove the attempt from the user when the SMS has been actually sent. In order to avoid the potential attack vector you describe, I added the `coolDown time` here https://github.com/vocdoni/blind-csp/blob/f/sms_queue/handlers/smshandler/mongodbstorage.go#L252

The coolDown time must be equal or higher than the queue TTL. This way a user can only ask for 1 SMS for a queue cycle.

With this in mind and taking into account my brain is starting to suffer from the last-minute programming, I'd love if you make a suggestion on the code.

Reply:
> also, implementing here the SmsThrottle sounds weird to me, and it leads to funny behaviour, since the channel is buffered (10), the throttle only comes into action after the channel is filled. in these conditions, the log lines (L120, L122, L125) are printed 5 seconds later = 10 * DefaultSMSthrottleTime(500ms) than the message that originated them.

I'm not sure I get your point. The throttle is happening even if the channel is buffered. It won't block the Queue processing (until the buffer is filled) but the `smsQueueController` func will only process 1 sms per `SmsThrottle` time.

Reply:
@mvdan your input on the SMS queue logic would be very useful :)

Reply:
> > also, implementing here the SmsThrottle sounds weird to me, and it leads to funny behaviour, since the channel is buffered (10), the throttle only comes into action after the channel is filled. in these conditions, the log lines (L120, L122, L125) are printed 5 seconds later = 10 * DefaultSMSthrottleTime(500ms) than the message that originated them.
> 
> I'm not sure I get your point. The throttle is happening even if the channel is buffered. It won't block the Queue processing (until the buffer is filled) but the `smsQueueController` func will only process 1 sms per `SmsThrottle` time.
correct, but it's not smsQueueController the one who sends the sms, it's `run()` 
https://github.com/vocdoni/blind-csp/blob/8d260e0239264730d4fd99af2b4efca8a113253f/handlers/smshandler/queue.go#L65
who blocks on the channel only after it's filled. so the first 10 sms are sent in a burst, then throtthled by smsQueueController.
also, the (max 10) retries of a failed sendChallenge are also sent in a burst, no throttling there.

Reply:
Then let's reduce the buffer of the channel to 1 item and the throttle will work :)